[String] optimize equalsTo/startsWith/endsWith

gharlan opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Tickets       | n/a
| License       | MIT
| Doc PR        | n/a

`ByteString::startsWith`: https://3v4l.org/7TkbQ
`UnicodeString::startsWith`: https://3v4l.org/gd91p
`UnicodeString::endsWith`: https://3v4l.org/Nhhv4

gharlan committed:
[String] optimize equalsTo/startsWith/endsWith

nicolas-grekas commented on src/Symfony/Component/String/UnicodeString.php:
```diff
@@ -97,11 +97,15 @@ public function endsWith($suffix): bool
+            return grapheme_strlen($this->string) - grapheme_strlen($suffix) === grapheme_strripos($this->string, $suffix);
+        }
+
+        return $suffix === grapheme_extract($this->string, \strlen($suffix), GRAPHEME_EXTR_MAXBYTES, -\strlen($suffix));
```
I'm not sure counting in bytes works with combining characters eg in NFD, this should differentiate an "e" from an "é".
Can you please add a test for this case?

gharlan replied:
added tests.

it works because of this behavior: https://3v4l.org/3CLS0

nicolas-grekas replied:
perfect, thank you

nicolas-grekas commented on src/Symfony/Component/String/UnicodeString.php:
```diff
@@ -332,7 +336,15 @@ public function startsWith($prefix): bool
+            return 0 === grapheme_stripos($this->string, $prefix);
+        }
+
+        return $prefix === grapheme_extract($this->string, \strlen($prefix), GRAPHEME_EXTR_MAXBYTES);
```
same as above: counting in bytes might not work

nicolas-grekas commented:
Please check tests, something is wrong :)
On Windows, the intl polyfills might not be up to the need and might need a fix...

gharlan commented:
> Please check tests, something is wrong :)

fixed.

> the intl polyfills might not be up to the need and might need a fix...

Added a workaround in this pr, and a fix for the polyfills: https://github.com/symfony/polyfill/pull/193

gharlan commented on src/Symfony/Component/String/UnicodeString.php:
```diff
@@ -332,7 +336,15 @@ public function startsWith($prefix): bool
+        }
+
+        if ($this->ignoreCase) {
+            return 0 === grapheme_stripos($this->string, $prefix);
```
Maybe this could be improved by using `grapheme_extract` and then `grapheme_stripos` against the extracted part.
@nicolas-grekas What do you think, is it worth it?
Are there graphemes with different byte length in lowercased and uppercased variant?
(the german `ß` in uppercase is `SS`, but this [is not matched by `*_stripos`](https://3v4l.org/Kf8rZ), so it does not matter here.)

nicolas-grekas replied:
a bench would tell... :)

gharlan replied:
looks good: https://3v4l.org/QeEdo

gharlan commented on src/Symfony/Component/String/UnicodeString.php:
```diff
@@ -332,7 +336,15 @@ public function startsWith($prefix): bool
+        }
+
+        if ($this->ignoreCase) {
+            return 0 === grapheme_stripos(grapheme_extract($this->string, \strlen($prefix), GRAPHEME_EXTR_MAXBYTES), $prefix);
```
hmm, maybe we could even use `mb_stripos` instead of `grapheme_stripos`, because grapheme_extract ensures that we do not get half grapheme clusters. Or is there a test case that would fail with `mb_stripos`?

gharlan replied:
the existing test cases are succeeding.

gharlan commented on src/Symfony/Component/String/UnicodeString.php:
```diff
@@ -118,7 +122,7 @@ public function equalsTo($string): bool
 
         if ('' !== $string && false !== $string && $this->ignoreCase) {
-            return grapheme_strlen($string) === grapheme_strlen($this->string) && 0 === grapheme_stripos($this->string, $string);
+            return \strlen($string) === \strlen($this->string) && 0 === mb_stripos($this->string, $string, 0, 'UTF-8');
```
@nicolas-grekas I've optimized case-insensitive equalsTo in CodePointString and UnicodeString.
Please take a look, if you see any problem with them.
(Again I'm not sure if there are any graphemes with different byte length in lowercased and uppercased variant.)

nicolas-grekas replied:
Please check https://unicode.org/faq/casemap_charprop.html it might have the answer.
I think there are cases when the byte length differs yes.

gharlan replied:
There are the one-to-one mappings, and the special mappings listed in ftp://ftp.unicode.org/Public/UCD/latest/ucd/SpecialCasing.txt.
Some of these special mappings have different byte length.
But the special mappings are not covered by `(grapheme|mb)_stripos` anyway.
So I think it is ok to compare the byte lengths here.

gharlan replied:
https://3v4l.org/BdUJs

nicolas-grekas commented:
Thank you @gharlan.
